### PR TITLE
Expose the scheduler affinity functions

### PR DESCRIPTION
**Pull request by faern**
I needed to set some affinity for my Rust threads so I added bindings to these functions.
Comments on the design is welcome.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
 license = "MIT"
 readme = "README.md"
-keywords = ["sched", "scheduler"]
+keywords = ["sched", "scheduler", "affinity"]
 exclude = ["Cargo.lock", ".gitignore"]
 
 [dependencies]
 errno = "*"
-libc = "*"
+libc = "0.2.*"
```
Comment (reem) [inline Cargo.toml]: 0.2 and 0.1 will mean the same thing

File: src/cpuset.rs [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    /// Get the raw pointer to the bitmask
+    /// Any modification of the `CpuSet` after this call might invalidate the pointer.
+    pub fn mask_ptr(&self) -> *const c_void {
+        &self.mask[0] as *const _ as *const c_void
+    }
+
+    /// Get a mutable raw pointer to the bitmask.
+    /// Any modification of the `CpuSet` after this call might invalidate the pointer.
+    pub fn mut_mask_ptr(&mut self) -> *mut c_void {
+        (&mut self.mask[0]) as *mut _ as *mut c_void
```
Comment (reem) [inline src/cpuset.rs]: `self.mask.as_mut_ptr() as *mut c_void` should do it

File: src/cpuset.rs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    /// Get the number of bytes in the mask.
+    /// Produces the same results as `CPU_ALLOC_SIZE`.
+    pub fn len(&self) -> usize {
+        (MASK_BITS / 8) * self.mask.len()
+    }
+
+    /// Get the raw pointer to the bitmask
+    /// Any modification of the `CpuSet` after this call might invalidate the pointer.
+    pub fn mask_ptr(&self) -> *const c_void {
+        &self.mask[0] as *const _ as *const c_void
```
Comment (reem) [inline src/cpuset.rs]: `self.mask.as_ptr() as *const c_void` here too

File: src/cpuset.rs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    /// Represent this `CpuSet` as an integer.
+    /// Will return an `Err` if the `CpuSet` is too large to be written to `T`
+    pub fn as_int<T: num::integer::Integer + num::traits::Zero>(&self) -> Result<T, ()> {
+        let src_size = self.len();
+        let out_size = mem::size_of::<T>();
+        if src_size > out_size {
+            Err(())
+        } else {
+            let mut mask: T = num::zero();
+            unsafe { ptr::copy(self.mask_ptr(), (&mut mask) as *mut _ as *mut c_void, src_size) }
```
Comment (reem) [inline src/cpuset.rs]: this is not safe, `T: num::integer::Integer` does not mean it is laid out like a CPU integer - this function should probably either return a concrete type or be marked `unsafe`

File: src/sched.rs [hunk 5]
```diff
@@ -59,3 +60,61 @@ pub fn get_policy(pid: i32) -> Result<Policy, ()> {
...
 }
+
+/// Set the cpu affinity for the current thread See `set_affinity()`.
+pub fn set_self_affinity(cpuset: CpuSet) -> Result<(), ()> {
+    set_affinity(0, cpuset)
+}
+
+/// Set the cpu affinity for a thread.
+pub fn set_affinity(pid: i32, cpuset: CpuSet) -> Result<(), ()> {
+    match unsafe { sched_setaffinity(pid, cpuset.len(), cpuset.mask_ptr()) } {
```
Comment (reem) [inline src/sched.rs]: I think that the unsafety of calling `sched_setaffinity` could be encapsulated in safe method on `CpuSet` that is just called here and on L#89

**Comment by reem** [comment]
Besides my comments this looks *great*, thank you very much! Unfortunately, this library isn't getting a lot of love internally so if you'd actually like to take it over completely, I would also be open to it.

**Comment by faern** [comment]
Many good comments. I implemented everything you recommended. I can take over the library if you want to, but there is nothing more I require from `sched.h` at the moment.

**Comment by reem** [comment]
We can just keep it then. Everything looks great now, can you squash to a smaller set of commits without merge commits and I'll merge? I just reconfigured travis, so hopefully it will run correctly then too.

**Comment by faern** [comment]
Ok, so I massaged the commit history into something cleaner. The two tests getting and setting the actual affinity fail. I guess Travis won't allow the tests to do system calls, which is reasonable :)
I don't know if there's any best practices around this? Should I maybe not test the system calls themselves? Mark them as ignored? Or just ignore the fact that Travis does not like my tests?

**Comment by reem** [comment]
The one thing to try is changing `sudo: false` to `sudo: required` and seeing if that changes anything (that way you run on legacy non-container infrastructure).

**Comment by faern** [comment]
Yup, `sudo: required` made it fly through all tests. I copied .travis.yml from your other branch, I guess you just merge that in.

**Comment by reem** [comment]
Ah, now there are conflicts. If you rebase your branch on master that should go away.

**Comment by reem** [comment]
Awesome!

**Pull request merged**